src/coreclr/jit/flowgraph.cpp

```diff
@@ -436,6 +436,308 @@ BasicBlock* Compiler::fgCreateGCPoll(GCPollType pollType, BasicBlock* block)
+                //
+
+                // The initialization check looks like this for JIT:
+                //
+                // *  JTRUE     void
+                // \--*  EQ        int
+                //    +--*  AND       int
+                //    |  +--*  IND       int
+                //    |  |  \--*  CNS_INT(h) long   0x.... const ptr
+                //    |  \--*  CNS_INT   int    1 (bit mask)
+                //    \--*  CNS_INT   int    1
+                //
+                // For NativeAOT it's:
+                //
+                // *  JTRUE     void
+                // \--*  EQ        int
+                //    +--*  IND       nint
+                //    |  \--*  ADD       long
+                //    |     +--*  CNS_INT(h) long   0x.... const ptr
+                //    |     \--*  CNS_INT   int    -8 (offset)
+                //    \--*  CNS_INT   int    0
+                //
+                assert(flagAddr.accessType == IAT_VALUE);
+
+                GenTree* cachedStaticBase = nullptr;
+                GenTree* isInitAdrNode;
+                if (isInitOffset != 0)
+                {
+                    assert(IsTargetAbi(CORINFO_NATIVEAOT_ABI));
+                    GenTree* baseAddr = gtNewIconHandleNode((size_t)flagAddr.addr, GTF_ICON_CONST_PTR);
+
+                    // Save it to a temp - we'll be using its value for the replacementNode.
+                    // This leads to some size savings on NativeAOT
+                    if ((staticBaseAddr.addr == flagAddr.addr) && (staticBaseAddr.accessType == flagAddr.accessType))
+                    {
+                        cachedStaticBase = fgInsertCommaFormTemp(&baseAddr);
+                    }
+
+                    // Don't fold ADD(CNS1, CNS2) here since the result won't be reloc-friendly for AOT
+                    isInitAdrNode = gtNewIndir(TYP_I_IMPL, gtNewOperNode(GT_ADD, TYP_I_IMPL, baseAddr,
+                                                                         gtNewIconNode(isInitOffset)));
+                    isInitAdrNode->gtFlags &= ~GTF_EXCEPT;
+                    isInitAdrNode->gtFlags |= (GTF_IND_NONFAULTING | GTF_IND_INVARIANT);
+                }
+                else
+                {
+                    assert(!IsTargetAbi(CORINFO_NATIVEAOT_ABI));
+                    isInitAdrNode = gtNewIndOfIconHandleNode(TYP_INT, (size_t)flagAddr.addr, GTF_ICON_CONST_PTR, true);
+                }
+
+                GenTree* isInitedValue;
+                if (!IsTargetAbi(CORINFO_NATIVEAOT_ABI))
+                {
+                    // In JIT we only check a single bit (see ClassInitFlags::INITIALIZED_FLAG)
+                    isInitAdrNode = gtNewOperNode(GT_AND, TYP_INT, isInitAdrNode, gtNewIconNode(1));
+                    isInitedValue = gtNewIconNode(1);
+                }
+                else
+                {
+                    // It was flipped to 0 for better codegen in https://github.com/dotnet/runtime/pull/83937
```

Comment:
Nit: It was not just for a better codegen. It was also for smaller data size. I think this comment can be omitted.